lib/elixir/lib/macro.ex

```diff
@@ -1967,10 +1967,53 @@ defmodule Macro do
   def quoted_literal?({:%{}, _, args}), do: quoted_literal?(args)
   def quoted_literal?({:{}, _, args}), do: quoted_literal?(args)
   def quoted_literal?({:__MODULE__, _, ctx}) when is_atom(ctx), do: true
+  def quoted_literal?({:<<>>, _, segments}), do: Enum.all?(segments, &quoted_bitstring_segment?/1)
   def quoted_literal?({left, right}), do: quoted_literal?(left) and quoted_literal?(right)
   def quoted_literal?(list) when is_list(list), do: :lists.all(&quoted_literal?/1, list)
   def quoted_literal?(term), do: is_atom(term) or is_number(term) or is_binary(term)
 
+  defp quoted_bitstring_segment?(term) when is_integer(term) or is_binary(term), do: true
+
+  defp quoted_bitstring_segment?({:"::", _, [term, modifier]})
+       when is_integer(term) or is_binary(term),
+       do: quoted_bitstring_modifier?(modifier)
+
+  defp quoted_bitstring_segment?(_), do: false
+
+  @bitstring_modifiers [
+    :integer,
+    :float,
+    :bits,
+    :bitstring,
+    :binary,
+    :bytes,
+    :utf8,
+    :utf16,
+    :utf32,
+    :signed,
+    :unsigned,
+    :big,
+    :little,
+    :native
+  ]
+  # cannot use in/2 guard in Macro module for bootstrapping issues
+  defp quoted_bitstring_modifier?({modifier, _, ctx})
+       when is_atom(ctx) or ctx == [],
+       do: Enum.member?(@bitstring_modifiers, modifier)
```

Comment:
Since we can't use the `in` at compile time, we can probably optimize this and keep it just as readable by doing something like this:

```suggestion
  # cannot use in/2 guard in Macro module for bootstrapping issues
  defp quoted_bitstring_modifier?({modifier, _, ctx})
       when is_atom(ctx) or ctx == [],
       do: bitstring_literal_type?(modifier)

  defp bitstring_literal_type?(:integer), do: true
  defp bitstring_literal_type?(:float), do: true
  # ...
  defp bitstring_literal_type?(_other), do: false
```

Reply:
I like this idea better indeed 💜 Will apply it

Reply:
Done in [443f76b](https://github.com/elixir-lang/elixir/pull/12270/commits/443f76b19e11d644c6833d5e15c0b417ecf06d81)

Reply:
Or would `case` be less noisy than `defp` in this case?
```elixir
case modifier do
  :integer -> true
  :float -> true
  ...
end
```